src/layers/learning/channelwise_fully_connected.cpp

```diff
@@ -34,6 +34,297 @@
+  auto* linearity_optimizer = static_cast<data_type_optimizer<TensorDataType>*>(layer.get_weights(0).get_optimizer());
+
+  if (linearity_optimizer == nullptr){
+    dc::MPIRootPrintStreamInfo() << "Weights optimizer null. Exiting ...." << std::endl;
+    return;
+  } 
+  auto& linearity_gradient = linearity_optimizer->get_gradient_buffer(
+      dst_scale, gradient_scale, true);
+
+  assert0(dc::tensor::View(*m_linearity_gradient, linearity_gradient.Buffer()));
+  m_linear_operator->backward_wrt_weight(layer.m_transpose,
+                                         dst_scale,
+                                         gradient_scale,
+                                         this->get_prev_activations(),
+                                         this->get_prev_error_signals(),
+                                         *m_linearity_gradient);
+
+  if(layer.m_has_bias){
+    auto* bias_optimizer = static_cast<data_type_optimizer<TensorDataType>*>(layer.get_weights(1).get_optimizer());
+    if (bias_optimizer == nullptr) return;
+    
+    auto& bias_gradient = bias_optimizer->get_gradient_buffer(
+        dst_scale, gradient_scale, true);
+
+    assert0(dc::tensor::View(*m_bias_gradient, bias_gradient.Buffer()));
+
+    m_linear_operator->backward_wrt_bias(gradient_scale,
+                                         dst_scale,
+                                         this->get_prev_error_signals(),
+                                         *m_bias_gradient);
+  }
+}
+
+
+template <typename TensorDataType, data_layout Layout, El::Device Device>
+dc::Shape
+channelwise_fully_connected_distconv_adapter<TensorDataType, Layout, Device>
+::get_activations_local_shape(int index) const{
+
+  // The default case assumes that the local is shape is the same as 
+  // the local shape of the first previous activations 
+
+  // Need to update such that the height and width dimensions match 
+  // match the output dimensions expected  
+
+  const auto &layer = dynamic_cast<const channelwise_fully_connected_layer
+    <TensorDataType, Layout, Device>&>(this->layer());
+
+  auto linearity_dims = layer.get_linearity_dims();
+
+  std::reverse(std::begin(linearity_dims), std::end(linearity_dims));
+  const auto output_shape = 
+    ::distconv::get_fc_output_local_tensor_shape(
+      this->get_prev_activations(), linearity_dims, layer.m_transpose);
+  return output_shape;
+}
+
+template <typename TensorDataType, data_layout Layout, El::Device Device>
+dc::Shape
+channelwise_fully_connected_distconv_adapter<TensorDataType, Layout, Device>
+::get_activations_shape(int index) const{
+  return data_type_distconv_adapter<TensorDataType>::get_activations_shape();
+}
+
+template <typename TensorDataType, data_layout Layout, El::Device Device>
+dc::Shape
+channelwise_fully_connected_distconv_adapter<TensorDataType, Layout, Device>
+::get_prev_activations_shape(int index) const{
+  return data_type_distconv_adapter<TensorDataType>::get_prev_activations_shape(0);
+}
+
```

Comment:
If we are just calling the base class implementation, do we need to override these functions?
```suggestion
```